File: core/export/markdown.py
# -*- coding: utf-8 -*-
"""Markdown 导出(带 YAML front matter,Obsidian 友好)。"""
from __future__ import absolute_import, unicode_literals

import io
import re


def _safe_quote(s):
    return (s or "").replace('"', '\\"')


def _fmt_date(iso_text):
    m = re.match(r"^(\d{4})-(\d{2})-(\d{2})", iso_text or "")
    return m.group(0) if m else (iso_text or "")


def export_markdown(session, analysis, response, out_path=None):
    """生成 Markdown 文档。返回文本。"""
    tags = analysis.tags or []
    quotes = analysis.golden_quotes or []
    fm = [
        "---",
        "title: %s" % _safe_quote(
            "与%s的深度对话" % session.peer),
        "date: %s" % _fmt_date(analysis.start_time),
        "peer: %s" % _safe_quote(session.peer),
        "depth_score: %d" % int(round(analysis.depth_score)),
        "tags: [%s]" % ", ".join(tags),
        "session_id: %s" % session.session_id,
        "---",
        "",
    ]
    body = [
        "## 对话摘要",
        "",
        analysis.summary,
        "",
        "## AI 第三方回应",
        "",
        response["text"],
        "",
        "## 金句",
        "",
    ]
    for q in quotes[:3]:
        body.append("- %s" % q["text"])
    body += ["", "## 完整对话", ""]
    for m in session.messages:
        sender = "我" if m.sender == "me" else m.sender
        ts = _fmt_date(m.timestamp) + " " + re.sub(
            r"^(\d{4}-\d{2}-\d{2})T(\d{2}:\d{2}).*", r"\2",
            m.timestamp or "")
        label = sender or "对方"
        body.append("### %s · %s" % (label, ts))
        body.append("")
        body.append(m.content)
        body.append("")
    text = "\n".join(fm + body)
    if out_path:
        with io.open(out_path, "w", encoding="utf-8", newline="\n") as f:
            f.write(text)
    return text

File: src/export/markdown.ts
import type { ExportInput, ExportOptions } from './types.js';
import { buildTimeInfo, formatClock } from './time.js';
import { peerLabel, scoreVisible } from './privacy.js';
import { buildTitle } from './title.js';

function yamlString(value: string): string {
  return JSON.stringify(value);
}

/** 渲染 Markdown 可迁移笔记(YAML front matter + 正文)。 */
export function renderMarkdown(input: ExportInput, opts: ExportOptions): string {
  const brand = opts.brand ?? 'DeepTalk';
  const title = buildTitle(input, opts);
  const a = input.analysis;
  const time = buildTimeInfo(a);

  const lines: string[] = [];

  // —— YAML front matter(便于 Obsidian 检索)——
  lines.push('---');
  lines.push(`title: ${yamlString(title)}`);
  lines.push(`date: ${a.start_time.slice(0, 10)}`);
  if (scoreVisible(opts)) {
    lines.push(`peer: ${yamlString(input.session.peer)}`);
    lines.push(`depth_score: ${a.depth_score}`);
  } else {
    lines.push(`peer: ${yamlString('匿名')}`);
  }
  lines.push(`tags: [${a.tags.map(yamlString).join(', ')}]`);
  lines.push(`session_id: ${input.session.session_id}`);
  lines.push(`segment_id: ${a.segment_id}`);
  if (a.model) lines.push(`model: ${yamlString(a.model)}`);
  lines.push(`source: ${yamlString(brand)}`);
  lines.push('---');
  lines.push('');

  // —— 正文 ——
  lines.push(`# ${title}`);
  lines.push('');
  lines.push(
    `> ${time.dateLabel} · ${time.rangeLabel} · 时长 ${time.durationLabel} · ${time.sinceLabel}`,
  );
  lines.push('');
  lines.push(`- 对话对象:${peerLabel(input, opts)}`);
  if (scoreVisible(opts)) lines.push(`- 深度评分:${a.depth_score}`);
  if (a.dimensions) {
    const d = a.dimensions;
    lines.push(
      `- 四维评分:情感 ${d.emotion} / 事件 ${d.event} / 连续 ${d.continuity} / 互动 ${d.interaction}`,
    );
  }
  lines.push(`- 标签:${a.tags.join('、') || '（无）'}`);
  lines.push('');

  lines.push('## 对话摘要');
  lines.push('');
  lines.push(a.summary);
  lines.push('');

  lines.push('## AI 第三方回应');
  lines.push('');
  lines.push(input.aiResponse ?? '（暂无）');
  lines.push('');

  lines.push('## 完整对话');
  lines.push('');
  for (const m of a.messages) {
    const who = m.sender === 'me' ? '我' : scoreVisible(opts) ? m.sender : '对方';
    lines.push(`**${who}** _(${formatClock(m.timestamp)})_: ${m.content}`);
    lines.push('');
  }

  return lines.join('\n');
}
